### PR TITLE
DTLS SRTP (RFC5764) support (adds `--enable-srtp`)

### PR DESCRIPTION
**Pull request by dgarske**
Used with DTLS to agree on profile for new real-time session keys, like what is used for WebRTC. Supports all recent SRTP profiles and client sending multiple.

This has completed interoperability testing with openssl and it builds and tests successfully against WebRTC.

This PR includes internal automated ongoing tests and an interop script `scripts/openssl_srtp.test`.

Examples:

```
./configure --enable-opensslall --enable-dtls --enable-srtp --enable-keying-material && make

./examples/client/client -u -x -g --srtp SRTP_AES128_CM_SHA1_80
./examples/server/server -u -d -g --srtp SRTP_AES128_CM_SHA1_80

openssl s_client -use_srtp SRTP_AES128_CM_SHA1_80 -port 11111 -dtls1_2
openssl s_server -use_srtp SRTP_AES128_CM_SHA1_80 -port 11111 -dtls1_2 -cert ./certs/server-cert.pem -key ./certs/server-key.pem
```

Then call `wolfSSL_export_dtls_srtp_keying_material ` or `SSL_export_keying_material` to get secret for real-time session.

### CODE REVIEW

**Review by lealem47 — APPROVED** [review]

File: src/internal.c [hunk 1]
```diff
@@ -485,18 +485,32 @@ static WC_INLINE int IsEncryptionOn(WOLFSSL* ssl, int isSend)
...
 #else
     return ssl->options.dtls;
 #endif
 }
-#endif /* DTLS || !WOLFSSL_NO_TLS12 */
+
+/* Secure Real-time Transport Protocol */
+/* If SRTP is not enabled returns the state of the dtls option.
+ * If SRTP is enabled returns dtls && !dtlsSrtpProfiles. */
+int IsDtlsNotSrtpMode(WOLFSSL* ssl)
```
Comment (SparkiDev) [inline src/internal.c:504]: Only appears to be used locally.
Reduce scope if possible (static WC_INLINE).

File: src/tls.c [hunk 2]
```diff
@@ -5351,6 +5351,184 @@ int TLSX_EncryptThenMac_Respond(WOLFSSL* ssl)
...
+            }
+        }
+        (void)profile_len;
+    }
+
+    if (ret != 0) {
+        WOLFSSL_MSG("SRP Profile not found!");
+        ssl->dtlsSrtpId = 0;
+        TLSX_UseSRTP_Free(srtp, ssl->heap);
+        /* not fatal, so return 0 */
```
Comment (SparkiDev) [inline src/tls.c:5463]: MEMORY_E not fatal?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/tls.c [hunk 3]
```diff
@@ -5351,6 +5351,189 @@ int TLSX_EncryptThenMac_Respond(WOLFSSL* ssl)
...
+                break;
+            }
+        }
+        (void)profile_len;
+    }
+
+    if (ret == 0 && ssl->dtlsSrtpId == 0) {
+        WOLFSSL_MSG("SRP Profile not found!");
+        /* not fatal, so return 0 */
+        ret = 0;
```
Comment (SparkiDev) [inline src/tls.c:5465]: ret == 0 from If

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/tls.c [hunk 4]
```diff
@@ -5351,6 +5351,165 @@ int TLSX_EncryptThenMac_Respond(WOLFSSL* ssl)
...
+    word16 offset = 0;
+#endif
+
+    if (length < OPAQUE16_LEN) {
+        return BUFFER_ERROR;
+    }
+
+    if (!isRequest) {
+#ifndef NO_WOLFSSL_CLIENT
+        ssl->dtlsSrtpProfile = ssl->ctx->dtlsSrtpProfile;
```
Comment (rizlik) [inline src/tls.c]: Here we should set `ssl->dtlsSrtpId` with the one selected by the server (after checking that is in the list we sent to him)
I used this and worked:
```
modified   src/tls.c
@@ -5421,7 +5421,20 @@ static int TLSX_UseSRTP_Parse(WOLFSSL* ssl, const byte* input, word16 length,
 
     if (!isRequest) {
 #ifndef NO_WOLFSSL_CLIENT
-        ssl->dtlsSrtpProfiles = ssl->ctx->dtlsSrtpProfiles;
+        /* skip length */
+        offset += OPAQUE16_LEN;
+        if (length - offset < OPAQUE16_LEN)
+            return BUFFER_ERROR;
+
+        ato16(input + offset, &profile_value);
+
+        /* check that the profile received was in the ones we support */
+        if ((ssl->dtlsSrtpProfiles & (1 << profile_value)) == 0) {
+            WOLFSSL_MSG("SRP profile sent by the server not supported!");
+            return 0;
+        }
+
+        ssl->dtlsSrtpId = profile_value;
         ret = 0; /* success */
 #endif
     }
```

File: src/tls.c [hunk 5]
```diff
@@ -5351,6 +5351,165 @@ int TLSX_EncryptThenMac_Respond(WOLFSSL* ssl)
...
+#endif
+    }
+#ifndef NO_WOLFSSL_SERVER
+    else {
+        /* total length, not include itself */
+        ato16(input, &profile_len);
+        offset += OPAQUE16_LEN;
+        /* parse remainder one profile at a time, looking for match in CTX */
+        for (i=offset; i<length; i+=OPAQUE16_LEN) {
+            ato16(input+i, &profile_value);
```
Comment (rizlik) [inline src/tls.c:5438]: are we accounting for the MKI field in this parsing?

**Review by rizlik — CHANGES_REQUESTED** [review]

File: src/ssl.c [hunk 6]
```diff
@@ -1277,23 +1277,38 @@ int wolfSSL_dtls_set_mtu(WOLFSSL* ssl, word16 newMtu)
...
+    const char* profile_str, word32 profile_str_len, unsigned long id)
 {
     int i;
     const WOLFSSL_SRTP_PROTECTION_PROFILE* profile = NULL;
     for (i=0;
          i<(int)(sizeof(gSrtpProfiles)/sizeof(WOLFSSL_SRTP_PROTECTION_PROFILE));
          i++) {
         if (profile_str != NULL) {
-            if (XSTRCMP(gSrtpProfiles[i].name, profile_str) == 0) {
+            if (XMEMCMP(gSrtpProfiles[i].name, profile_str, profile_str_len)
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/ssl.c:1310]: Yes, this is not null terminated and I did not want to make a copy of the string just to be able to use strncmp. The original code supported only a single SRTP profile, but with the support for colon ":" separated lists the memcmp is better.

File: src/ssl.c [hunk 7]
```diff
@@ -1305,64 +1320,112 @@ static const WOLFSSL_SRTP_PROTECTION_PROFILE* FindDtlsSrtpProfile(
...
+    
+    *id = 0; /* reset destination ID's */
+
+    if (profile_str == NULL) {
+        return WOLFSSL_FAILURE;
+    }
+
+    /* loop on end of line or colon ":" */
+    next = profile_str;
+    length = (word32)XSTRLEN(profile_str);
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/ssl.c:1339]: `length` is used below to make sure the the string is not larger than `current_length`.  ```if (current_length < length) length = current_length;```

File: src/tls.c [hunk 8]
```diff
@@ -5351,6 +5351,165 @@ int TLSX_EncryptThenMac_Respond(WOLFSSL* ssl)
...
+
+    extension = TLSX_Find(*extensions, TLSX_USE_SRTP);
+    if (extension == NULL) {
+        ret = TLSX_Push(extensions, TLSX_USE_SRTP, (void*)srtp, heap);
+        if (ret != 0) {
+            TLSX_UseSRTP_Free(srtp, heap);
+        }
+    }
+
+    return ret;
```
[the remark this answers is not in the document]
Comment (dgarske) [reply src/tls.c:5530]: No, this just means it was already added to the TLS extensions list. It is free'd elsewhere.

File: src/tls.c (hunk 5 above)
Comment (dgarske) [reply src/tls.c:5449]: No, the MKI is a byte that comes at the end. Always zero on send and ignored on parse. None of the implementations I found used MKI, so I did not add support for it.

**Review by SparkiDev — APPROVED** [review]

File: src/ssl.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (rizlik) [reply src/ssl.c:1310]: Oh, I see now why we can't use `XSTRCMP`. But using `XMEMCMP` leads us to compare past the `gStrpProfiles[i].name` buffer if `profile_str_len` is bigger than `strlen(gStrpProfiles[i].name)`. Indeed -fsanite=address catches this. I tried to fix this in 86ba0ef6439100fa5054ba4cd1ae6a1a6a997b00 but at the cost of one extra `strlen`

File: src/ssl.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (rizlik) [reply src/ssl.c:1339]: I still can't understand this. 
`length` is used in this check (`if (current_length < length) length = current_length;`), but this check only updates `length` itself. Besides that `length` is never used. From what I see we can remove both `length` and the `check`. Am I missing something?

File: src/tls.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (rizlik) [reply src/tls.c:5530]: I was thinking at his execution flow:
```
srtp = TLSX_UseSRTP_New(); <- new allocation
extension = TLSX_Find(); <- this returns NULL
```
we go out of the function without freeing the just allocated `srtp`

File: tests/test-dtls-srtp.conf [hunk 9]
```diff
@@ -0,0 +1,192 @@
+# server DTLSv1.2 SRTP default profile
+-u
+-d
+-l ECDHE-RSA-AES256-GCM-SHA384
```
Comment (rizlik) [inline tests/test-dtls-srtp.conf:4]: @dgarske 
It looks like we have to explicitly set the cipher suite in the test. I just pick these, is there a better alternative? What if the ciphersuite isn't compiled in?

Comment (dgarske) [reply tests/test-dtls-srtp.conf:4]: Correct, the suites tests uses the cipher suite name to determine if it is compiled in or not and gates on it. I don't see value is choosing more than one. More importantly testing the exported key material for each profile is correct.

File: src/ssl.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (dgarske) [reply src/ssl.c:1339]: I stole this logic from the `static int CheckcipherList(const char* list)` function. The purpose is to make sure we aren't checking past the min length of each string.

File: src/ssl.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgarske) [reply src/ssl.c:1310]: Thanks for fixing. You cannot use `strlen`, it should be `XSTRLEN()`. Also you will get a type warning. Please cast.
`word32 srtp_profile_len = (word32)XSTRLEN(gSrtpProfiles[i].name);` Don't use continue, instead add the check before the XMEMCMP.

File: src/tls.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (dgarske) [reply src/tls.c:5530]: Yes this is wrong.

File: wolfssl/test.h [hunk 10]
```diff
@@ -629,6 +643,82 @@ err_sys_with_errno(const char* msg)
...
+ * @ekm: where to store the shared buffer pointer [out]
+ * @size: size of the shared buffer returned [out]
+ *
+ * This function wait that the other peer calls strp_helper_set_ekm() and then
+ * store the buffer pointer/size in @ekm and @size.
+ */
+static WC_INLINE void srtp_helper_get_ekm(srtp_test_helper *srtp,
+                                          uint8_t **ekm, size_t *size)
+{
+#if defined(_POSIX_THREADS) && !defined(__MINGW32__)
```
Comment (dgarske) [inline wolfssl/test.h:670]: As discussed please test with the non pthreads case to make sure the test passes... even if it is skipped. Thanks

**Review by dgarske — COMMENTED** [review]
Thanks @rizlik for all your work on this PR! Over to @SparkiDev for final review/merge. If you'd like me to squash this into just a few commits let me know.

**Review by rizlik — APPROVED** [review]

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**